### DISCUSSION TITLE
Adding more options for volume

### DISCUSSION THREAD

**Discussion by nicymike** [AlmasB/FXGL#1180] [Ideas]
Hey!

1. What do you think about adding a new setting to audio for master volume.
Many games have three (or sometimes more) volume sliders: master, music, sound.
The volume for playing audio is then calculated by:
soundVolumeForPlay = master * sound
musicVolumeForPlay = master * music

2. What do you think about adding the possibility to play a sound with a different volume eg. adding a new optional parameter to the methods for playing a sound or music.
In my game I have certain sounds for ambience which should be played with lower volume.
What I am currently doing with FXGL is: 1. Lower sound volume, 2. Play sound, 3. Raise sound volume.

**Comment by AlmasB** [comment]
Hi,

1) As far as I can remember, existing volume for sounds and music are the master volume for their respective types.

2) Each sound / music object can be applied with its own volume. Try:

```
Sound sound = getAssetLoader().loadSound(...);
sound.setXXXX
```

Let me know if this works.

**Comment by nicymike** [comment]
Thank you for your answers.

1. Yes, but I think about a separate volume setting which effects both types simultaneously. In my game I call the third setting the main volume which exists in addition to sound and music and will be used as a multiplier (as described in my original post).
![image](https://user-images.githubusercontent.com/19415000/173122351-4cd10441-a1b1-4d7c-8a3c-577787353c7d.png)

2. My Eclipse does not get the connection to the class Audio and therefore does not find the methods. Do you have any ideas what I am doing wrong?
![image](https://user-images.githubusercontent.com/19415000/173122269-0ff95295-c8b4-4c96-9a5b-bb25a268ab09.png)

**Reply by AlmasB** [reply]
1. Given it is relatively easy to create a new property to bind the two existing ones, I am not sure whether it is needed to be on the engine side.

2. That's a good point. It seems the implementation is hidden from public API. This might have been necessary to hide cross-platform support (particularly on mobile), but I don't see why we couldn't open access to `getAudio()`. From the list you will note the method is actually there, just hidden in `$fxgl_core`. I'm happy to turn this into an issue, so we can track it and fix it eventually.

**Comment by nicymike** [comment]
1. Thank you, I am ok with that.

2. I would be glad if you could open an access to audio.

**Reply by AlmasB** [reply]
Done in #1187

**Comment by AlmasB** [comment]
@nicymike @chengenzhao 

This is now public:

```
var sound = getAssetLoader().loadSound("file.wav");
var audio = sound.getAudio();

getAudioPlayer().playSound(sound);

// by default audio player will manage the volume, but we can manually override
audio.setVolume(volume);
```

**Reply by chengenzhao** [reply]
It works
but there is a small problem
developers have to play sound/music first
then set the volume
it doesnt work if the order is reversed
like 
```java
audio.setVolume(volume); 
getAudioPlayer().playSound(sound); // it won't work
```
